internal/browser/browser.go

```diff
@@ -0,0 +1,376 @@
+		if t.lastIteration.rowsFinished {
+			return t, nil
+		}
+		if t.lastIteration.Iterating {
+			return t, nil
+		}
+		go t.lastIteration.Iterate(50)
+		return t, t.lastIteration.IterateCmd(50 * time.Millisecond)
+	case tea.KeyMsg:
+		switch m.Type {
+		case tea.KeyTab:
+			t.keyboardFocus = !t.keyboardFocus
+			tuiutil.Faint = !tuiutil.Faint
+			return t, nil
+		case tea.KeyCtrlC:
+			if t.lastIteration != nil {
+				t.lastIteration.rows.Close()
+			}
+			return t, nil
+		}
+		if !t.keyboardFocus {
+			return t, nil
+		}
+	case tea.MouseMsg:
+		// disable all mouse events
+		return t, nil
+	case tea.WindowSizeMsg:
+		if m.Height > 0 {
+			m.Height += -2 // footer, header height offset
+		}
+		msg = m
+	}
+	oldYOffset := t.termdbmsTable.Viewport.YOffset + t.termdbmsTable.GetRow()
+	tmp, cmd := t.termdbmsTable.Update(msg)
+	t.termdbmsTable = tmp.(viewer.TuiModel)
+	newYOffset := t.termdbmsTable.Viewport.YOffset + t.termdbmsTable.GetRow()
+	if t.lastIteration != nil {
+		userOnLastPage := newYOffset > t.lastIteration.totalProcessedLines-t.termdbmsTable.Viewport.Height
+		if newYOffset > oldYOffset {
+			if userOnLastPage {
+				cmd = tea.Batch(cmd, func() tea.Msg {
+					return FetchMoreRowsMsg{}
+				})
+			}
+		}
+
+	}
+	return t, cmd
+}
+
+type SqlIterator struct {
+	rows                *sql.Rows
+	resultPipe          chan []interface{}
+	rowsFinished        bool
+	totalProcessedLines int
+	columnNames         []string // cannot access these after rows.Close(), hence save them
+	Iterating           bool
+}
+
+func NewSqlIterator(maxIterationCount int, rows *sql.Rows) *SqlIterator {
+	var si SqlIterator
+	si.rows = rows
+	si.columnNames, _ = rows.Columns()
+	si.resultPipe = make(chan []interface{}, maxIterationCount+1)
+	go si.Iterate(maxIterationCount)
+	return &si
+}
+
+func (si *SqlIterator) Iterate(maxIterationCount int) {
+	si.Iterating = true
+	var i int
+	for i = 0; si.rows.Next() && i < maxIterationCount; i++ { // each row of the table
+		// golang wizardry
+		columns := make([]interface{}, len(si.columnNames))
+		columnPointers := make([]interface{}, len(si.columnNames))
+		// init interface array
+		for i := range columns {
+			columnPointers[i] = &columns[i]
+		}
+		if err := si.rows.Scan(columnPointers...); err != nil {
+			break
+		}
+		si.resultPipe <- columnPointers
+	}
+	if i != maxIterationCount { // means query finished and there will be no more results
+		close(si.resultPipe)
+	}
+	si.Iterating = false
+}
+
+type NewRowsMessage [][]interface{}
+
+func (si *SqlIterator) IterateCmd(timeline time.Duration) func() tea.Msg {
+	return func() tea.Msg {
+		var newRows [][]interface{}
+		var timeout bool
+		for {
+			select {
+			case row := <-si.resultPipe:
+				if row == nil {
+					si.rowsFinished = true
+					return NewRowsMessage(newRows)
+				}
+				si.totalProcessedLines++
+				newRows = append(newRows, row)
+			case <-time.After(timeline):
+				timeout = true
+				break
+			}
+			if timeout {
+				break
+			}
```

Comment:
You can set a label to the for loop and `break loop` to simplify this logic:
```
loop:
for {
   // ...
     case <- t.C:
         break loop
   // ...
}
```

Reply:
https://github.com/hazelcast/hazelcast-commandline-client/pull/65/commits/64422a75715ba9e8683e02b9908ceb611c5f4832